src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -390,6 +390,82 @@ def test_all_endpoints_have_valid_composition(self):
+        for endpoint_id, endpoint in self.endpoints.items():
+            if Clusters.PowerSource not in endpoint:
+                continue
+            if Clusters.PowerSource.Attributes.EndpointList not in endpoint[Clusters.PowerSource]:
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem=f'Did not find {attribute_string} on {location.as_cluster_string(self.cluster_mapper)}', spec_location="EndpointList Attribute")
+                success = False
+            endpoint_list = endpoint[Clusters.PowerSource][Clusters.PowerSource.Attributes.EndpointList]
+            if endpoint_list and not all(id in self.endpoints.keys() for id in endpoint_list):
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem=f'{attribute_string} lists a non-existent endpoint', spec_location="EndpointList Attribute")
+                success = False
+
+        self.print_step(3, "Verify that all Bridged Node endpoint lists are correct")
+        device_types = {}
+        parts_list = {}
+        for endpoint_id, endpoint in self.endpoints.items():
+            if Clusters.PowerSource not in endpoint or Clusters.PowerSource.Attributes.EndpointList not in endpoint[Clusters.PowerSource]:
+                continue
+            if Clusters.Descriptor not in endpoint or Clusters.Descriptor.Attributes.PartsList not in endpoint[Clusters.Descriptor] or Clusters.Descriptor.Attributes.DeviceTypeList not in endpoint[Clusters.Descriptor]:
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=Clusters.Descriptor.id,
+                                                 attribute_id=Clusters.Descriptor.Attributes.PartsList.id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem='Missing descriptor cluster PartsList or DeviceTypeList', spec_location="PartsList Attribute")
+                success = False
+            device_types[endpoint_id] = [i.deviceType for i in endpoint[Clusters.Descriptor]
+                                         [Clusters.Descriptor.Attributes.DeviceTypeList]]
+            parts_list[endpoint_id] = endpoint[Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList]
+
+        bridged_nodes = [id for (id, dev_type) in device_types.items() if 0x13 in dev_type]
+
+        for endpoint_id in bridged_nodes:
+            if Clusters.PowerSource not in self.endpoints[endpoint_id]:
+                continue
+            desired_endpoint_list = parts_list[endpoint_id].copy()
+            desired_endpoint_list.append(endpoint_id)
+            desired_endpoint_list.sort()
```

Comment:
Should we use a set? Or are we worried about diplication and a list is more correct (if so maybe a comment would be good)

Reply:
yes, list for duplication. Added a comment.